### DISCUSSION TITLE
library dependencies

### DISCUSSION THREAD

**Discussion by Tocio88** [jnsbyr/arduino-samd21lpe#1] [Q&A]
Hi, I'm interested to use your library to get the samd21 in low power mode for a project at university on the Greenland ice cap. I have downloaded also the Arduino Helper library but I cannot make the program to compile due to all dependencies.
Could you give me a help to install all dependencies and library needed for make it run on the Arduino IDE? Thanks

**Comment by jnsbyr** [comment]
Hi Tocio88,

this library only depends on the SDK for the SAMD21 and I do not remember explicitly downloading the Arduino Helper library. Please try this:

- start with installing Arduino IDE 1.8.X
- use the board manager to install "Seed SAMD Board" 1.8.3 (this will install the Seed Studio SDK for the SAMD21)
- manually apply the "patches" to board.txt (C:\Users\<Username>\AppData\Local\Arduino15\packages\Seeeduino\hardware\samd\1.8.3) and startup.c (C:\Users\<Username>\AppData\Local\Arduino15\packages\Seeeduino\hardware\samd\1.8.3\cores\arduino)
- have a look at https://wiki.seeedstudio.com/Seeeduino-XIAO/ for extra tips regarding the Arduino IDE and the SAMD21 XIAO board

Even without the patching you should be able to build the library after selecting a SAMD21 board.

After building the library, open one of the examples - it should also build successful.

There are some uncommitted changes for RealTimeClock and TimerCounter that will follow soon.

Alternatively or additionally you should consider using VS Code as IDE. If this is of interest for you I could also commit them.

Finally there is a project I haven't found the time to document that uses the SAMD21, a solar panel, an energy harvesting chip, a very small lipo battery and a 433 MHz transmitter using this library - only the temp/hygro sensor is still waiting to be connected. The prototype is on my desk and running continuously since May mostly at very low ambient light (less than 100 lux) after initial charging.

Let me know if this helped to get it working or if you need more info.

**Comment by Tocio88** [comment]
Hi Jens,

Many thanks for your reply!
Last two days and nights I did diving into your library and tracking every
single error and dependencies and at end I managed to get it to work with
lot of googling and forums, and it compiles your simple EIC program.
Indeed I’m building this new type of simple weather station to track snow
melt on the Greenland Ice cap, and it made with a TTL sonic height sensor,
samd21 (I’m using an arduino zero for now but it is consuming too much I
found out so I’ll switch to other boards like sparkfun ones with less
component on it that eat milliamps for nothing), RTC, SPI FRAM memory, I2C
ublox gps for time sync and position and UART iridium satellite
communication, 20Ah lipo battery and 3W solar panel. Everything packed
inside a small box and as lightweight as possible.
I wrote already the program based on the RTCzero library, and was working
until I found out that the rtc.stanby() wasn’t making the samd sleeping.
Google helped me out to find that I had to disable the systemTicks and with
that I see the system sleeping but I’m still consuming too much power like
35mA. So I came across your library and started to dive into it.

What also I noticed from my previous days tests that I cannot have both
libraries (your and RTCzero) due to the ISR handler double function
declaration… So I was thinking to use your library to make repetitive
interrupts to wake up the board to make measurements, store data and send
data, and using an external RTC clock with physical interrupt to get
timestamp to save my data.

Am I correct or am I missing something?

This is kind new programming for me. Till now I have used research grade
Campbell data loggers for all our weather stations but for this specific
type of weather station we need something cheap, tiny and less power
consumption that the others.

But till now it has been an interesting 2 weeks of getting crazy hitting my
head lot of times and fixing all problems/bugs one by one, so it’s nice,
and right now I have around 1000line of code running at my desk just to
test the routines if everything works.


I’m also very interested in your prototype project, which chips are you
using to maximise the energy harvest?

Best regards,
Giorgio

**Comment by jnsbyr** [comment]
Hi Giogio,

I think you are on the right track for your project. Sound very interesting to me, 

Part of the problem with very low power consumption is the total number of components and the way they are wired. Little details can change the minimum power requirements significantly in both directions (see comments to my mods of XIAO board in the README). On one hand you probably don't want to design your own board, but on the over hand it will be hard to optimize if you don't.

As you already noticed, very low power consumption projects need a different approach how to code. An MCU typically does not come with an OS, so you are responsible for everything. RTC wakeup is a what I use in my project, too, but the SAMD21 RTC is good enough for me, as I just need a fixed 3 min period. Only if you have to do something at a specific time with very high precision, e.g. at 12:00:00.000Z, than I would look for a good external RTC. In all other cases the additional chip is a waste of power. The SAMD21 RTC can also be used in calendar mode, but you would need to extend my lib a little bit for this by improving on RTCzero. If you have GPS you can sync the calendar time once in a while to stay current.

Anyway: You obviously save the most power by turning everything off that is not needed.

Mixing the various low power libs will not work as they often use the same resources and try to to apply conflicting configurations. Most of the popular libs are practical for general use but they don't deliver max. power saving. My lib ties to take power saving a step further. But to make the most of it, please acquaint yourself with the clock system of the SAMD21 if you have not already done so. E.g. it is very important to understand that lowering the clock of a peripheral will save power but will also increase the access time and that will also cost power. The specific use case decides which configuration is best.

I created the lib after some initial tests. Have a look at the [README.md](https://github.com/jnsbyr/arduino-lowpowertest230203/blob/main/README.md) of my test project. It compares several chips and provides some consumption values for the SAMD21.

Regarding my solar DHT project, I use a 68x37 mm panel with a max. output of 5V and 0.3W (using a larger panel would make things too easy). The LIPO has 50mAh only. The energy harvesting is handled by a TI bq25570. The TI chip fills a capacitor from the panel and when the voltage of the capacitor is high enough the chips starts to charge the LIPO and/or power the output. To make it a little easier, I used a CJMCU-2557 board that comes with a super cap. The disadvantage of this board is that the output voltage is fixed to 2.5 V by soldered resistors. The code of my solar DHT project should give you a good idea how to make use of the lib for a non-trivial use case. I will try to post the code within the next 10 days, but no promises.

Kind regards,
Jens

**Reply by Tocio88** [reply]
Hi Jens,
I made some progress regarding power consumption. I ordered the Sparkfun atsamd21 dev Breakout board, and with the same program without using yet your library, the power consumption went down from 35mA of the arduino zero to 1mA in sleep.
That's already a huge step forward in the right direction but i still want to use your library but as soon i change the board to the sparkfun, i get tons of messages regarding missing declared variables as ULONG_MAX, UCHAR_MAX etc in the Arduino IDE...
So i need to understand why as well to try to fix it.
Regarding the RTC, i just need that wakes up at regular interval and do the sensor reading or other things, so the absolute time isn't necessary because i'm time stamping everything.
We looked into your energy harvesting chips and we will give it a try definitely but for that would be nice if i can go further with lowering the power consumption.
Kind regards,
Giorgio

**Reply by Tocio88** [reply]
reply: i solved the problems with the dependencies! Now i can start to try your library in the meantime you have time to post your code so i can study it :)

**Reply by jnsbyr** [reply]
I am currently picking up all the theads that I had to let unattended since end of September. 

The source of the project I have referred to is now online as [Solar DHT](https://github.com/jnsbyr/samd21-solardht), but still needs a lot of documentation.

To minimize power consumption all parts of the application have to be considered. As a result I had to make significant changes to almost all projects that Solar DHT depends on. Currently the project will not build successfully with the publicly available versions of the [ISM transmitter](https://github.com/nopnop2002/Arduino-SI4432). The changes for the [DHT sensor](https://github.com/RobTillaart/SHT2x) have been merged with the main project a few hours ago.

All these details will probably not be decisive for you. Have a look at the code from Solar DHT, especially how the functions of this library are used (Arduino main loop, timers, SAMD21 generic clocks, sleep modes).
